layout: Fix ordering of padding, border, and margin in inline BiDi

d-desiatkin opened the pull request:
<!-- Please describe your changes on the following line: -->
Previously visual order of inline box elements was not maintained in Layout 2020.
https://drafts.csswg.org/css-writing-modes-4/#bidi-fragment-boxes
2.4.5.2. Box Model of Reordering-induced Box Fragments

This commit adds naive implementation and allows following WPT tests to pass:
https://wpt.fyi/results/css/CSS2/box?label=master&product=chrome%5Bexperimental%5D&product=edge%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=servo&aligned

Probably we still can have some problems. Especially related to vertical writing modes. 
That issues I would like to solve in the separate commits.

Manual testing on local machine on following folder provided following results:
tests/wpt/tests/css/CSS2
Ran 6261 tests finished in 1425.6 seconds.
  • 8 ran as expected.
  • 65 tests failed unexpectedly
  • 35 tests passed unexpectedly
  
That is fairly unexpected. Will see what github actions will provide.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

d-desiatkin committed:
[fix](inline flow): Create PBM items in inline context in propper visual order

Signed-off-by: Desiatkin Dmitrii <d.desyatkin@innopolis.university>

nicoburns commented on components/layout_2020/flow/inline/mod.rs:
```diff
@@ -92,6 +92,7 @@ use serde::Serialize;
 use servo_arc::Arc;
 use style::computed_values::text_wrap_mode::T as TextWrapMode;
 use style::computed_values::vertical_align::T as VerticalAlign;
+use style::computed_values::direction::T as ComputedCSSDircetion;
```
You have a typo here. `Dircetion` instead of `Direction`

d-desiatkin replied:
TY. Will fix

d-desiatkin replied:
Fixed. Want to ask what is the policy here. I like to solve small typos by git amend and rewriting the original commit with forced push. Is it okay here? Or would you prefer to have separate commit?

nicoburns replied:
Either is fine! CI deals with force pushes to branches, and we squash merge so extra commits disappear anyway.

Loirooriol commented on components/layout_2020/flow/inline/mod.rs:
```diff
@@ -727,15 +728,27 @@ impl<'layout_dta> InlineFormattingContextLayout<'layout_dta> {
-                ));
+                .push(match css_direction {
+                        ComputedCSSDirection::Ltr => {
+                            LineItem::LeftInlineBoxPaddingBorderMargin(inline_box.identifier)
```
I think these should be renamed to `InlineStartBoxPaddingBorderMargin` and `InlineEndBoxPaddingBorderMargin` or such, then you wouldn't need to check the direction here.

Instead, https://github.com/servo/servo/blob/bd37d5496d747c7106b2d8d73dca1212421d82b4/components/layout_2020/flow/inline/line.rs#L379-L389 should just swap if the direction doesn't match the containing block.

d-desiatkin replied:
I don't think that it is proper change. Because if we change order of items in inline_items container. Then inline boxes borders could potentially participate in general procedure of text runs and inline items bidi reordering.
I believe that we will need it in the future to pass this test.
![image](https://github.com/user-attachments/assets/4cbc425e-f4ef-46c7-8e64-17c45af563b1)

d-desiatkin replied:
Currently there are multiple problems why this test fail. Issues with white space collapsing. Issues with inline boxes borders reordering. I need to investigate but my current idea is that borders should also participate in reordering procedure.

d-desiatkin replied:
So can we have some talk regarding this issue? 
I will need to prepare for such talk. Will reread documentation once again and make some notes.

I understand that in the Draft authors explicitly talk about fragments. And initially I have idea of implementing it near the place that you suggest. TBH I haven't find exactly the place that you stated in comment.

So we just need to understand will boxes ever participate in reordering process together with text_runs

d-desiatkin replied:
@Loirooriol

d-desiatkin replied:
![image](https://github.com/user-attachments/assets/17df768e-f813-423b-83ba-9af8ec0aa449)

d-desiatkin replied:
So currently there is some confusion in specs I believe. It would be cool if CSS committee can decide where should we apply bidi reordering and clearly state that. Theoretically it could be done on fragments level purely. Or we need more clearly define 2 separate procedures.
I.E text_runs bidi reordering and box fragments reordering as a separate term.

d-desiatkin replied:
@Loirooriol Look into Zulip.
Maybe I should to the opposite and extract this from fragment level to inline_items level?

d-desiatkin replied:
I can experiment tomorrow by adding control symbols to each inline item. Instead of simple copy of the last bidi level. And look into general behavior

Loirooriol replied:
> I don't think that it is proper change. Because if we change order of items in inline_items container

I'm confused, I didn't mean to reorder anything? I will push the change that I meant.

Loirooriol committed:
Rename left/right pbm terms in inline layout to logical nomenclature

 - LineItem::LeftInlineBoxPaddingBorderMargin(_) is renamed to
   LineItem::InlineStartBoxPaddingBorderMargin(_)
 - LineItem::RightInlineBoxPaddingBorderMargin(_) is renamed to
   LineItem::InlineEndBoxPaddingBorderMargin(_)
 - LineLayoutInlineContainerFlags::HAD_LEFT_PBM is renamed to
   LineLayoutInlineContainerFlags::HAD_INLINE_START_PBM
 - LineLayoutInlineContainerFlags::HAD_RIGHT_PBM is renamed to
   LineLayoutInlineContainerFlags::HAD_INLINE_END_PBM

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

Loirooriol committed:
Update test expectations

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

Loirooriol commented:
> These changes fix https://github.com/servo/servo/issues/6092 https://github.com/servo/servo/issues/7681 https://github.com/servo/servo/issues/14750

Please don't claim to fix layout 2013 specific bugs that have already been fixed in layout 2020.
